src/Data/Primitive/Contiguous.hs

```diff
@@ -445,6 +450,52 @@ foldlM' f z0 arr = go 0 z0
       | otherwise = return acc1
 {-# INLINABLE foldlM' #-}
 
+filter :: (Contiguous arr, Element arr a)
+  => (a -> Bool)
+  -> arr a
+  -> arr a
+filter p arr = ifilter (\_ a -> p a) arr
+
+ifilter :: (Contiguous arr, Element arr a)
+  => (Int -> a -> Bool)
+  -> arr a
+  -> arr a
+ifilter p arr = runST $ do
+  marr :: MutablePrimArray s Word8 <- newPrimArray sz
+  let go1 :: Int -> Int -> ST s Int
+      go1 !ix !numTrue = if ix < sz
+        then do
+          atIx <- indexM arr ix
+          let !keep = p ix atIx
+          let !keepTag = I# (dataToTag# keep)
+          writePrimArray marr ix (fromIntegral keepTag)
+          go1 (ix + 1) (numTrue + keepTag)
+        else pure numTrue
+  numTrue <- go1 0 0
+  if numTrue == sz
+    then pure arr
+    else do
+      marrTrues <- new numTrue
+      let go2 !ixSrc !ixDst = if ixSrc < sz && ixDst < numTrue
```

Comment:
hmm... I had not considered that `ixDst < numTrue` could be used to help short-circuit the computation, but it seems like this predicate is better than `ixSrc < sz` since the former always finishes sooner or at the same time as the latter. However, you can go one step further and just remove the `ixSrc < sz` check entirely since `ixDst < numTrue ==> ixSrc < sz`.